Added autolink typing support

jacekbogdanski opened the pull request:
## What is the purpose of this pull request?

New feature

## Does your PR contain necessary tests?

All patches which change the editor code must include tests. You can always read more
on [PR testing](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_contributing_code-section-tests),
[how to set the testing environment](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests) and
[how to create tests](https://docs.ckeditor.com/ckeditor4/docs/#!/guide/dev_tests-section-creating-your-own-test)
in the official CKEditor documentation.

### This PR contains

- [x] Unit tests
- [x] Manual tests

## What changes did you make?

I used `textmatch` plugin to provide `autolink` typing support. Pattern matching is a little complicated but it's required if we want to utilize the same URL/Email regex for both pastings from clipboard and typing.

Closes #1815

Comandeer commented on plugins/autolink/plugin.js:
```diff
@@ -64,6 +115,55 @@
+	 * @member CKEDITOR.config
+	 */
+	CKEDITOR.config.autolink_emailRegex = /^[a-zA-Z0-9.!#$%&'*+\/=?^_`{|}~-]+@[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?(?:\.[a-zA-Z0-9](?:[a-zA-Z0-9-]{0,61}[a-zA-Z0-9])?)*$/g;
+	// Regex by (https://www.w3.org/TR/html5/forms.html#valid-e-mail-address).
```
Actually it's no longer in HTML 5.x, now it's only in [HTML LS](https://html.spec.whatwg.org/#e-mail-state-(type=email)).

jacekbogdanski commented:
I refactored code to expose `autolink_commitKeystrokes` configuration property and removed redundant keystrokes. With the current implementation, a user can decide which keystrokes use to accept link completion. Defaults are `enter` and `space`.

I also changed undo behavior - now it requires 3 undo steps to remove a link:
* undo commit keystroke if it creates undo step like `space` or `enter`
* undo link
* undo link text

Comandeer commented on plugins/autolink/plugin.js:
```diff
@@ -6,45 +6,75 @@
 
-				// If we found "<" it means that most likely there's some tag and we don't want to touch it.
-				if ( data.indexOf( '<' ) > -1 ) {
+					var matched = CKEDITOR.plugins.textMatch.match( editor.getSelection().getRanges()[ 0 ], matchCallback );
```
It could look something like that:

```javascript
function autolink( editor, condition ) {
	var matched = CKEDITOR.plugins.textMatch.match( editor.getSelection().getRanges()[ 0 ], matchCallback );

		if ( !condition( matched ) ) {
			return;
		}

		insertLink( matched );
}

editor.on( 'paste', function( evt ) {
	// Pasted text may be encoded. Decode it so we will be able to compare its length with match.
	var text = CKEDITOR.tools.htmlDecode( evt.data.dataValue );

	// Attach afterPaste event here to avoid race condition between events.
	editor.once( 'afterPaste', function() {
		autolink( editor, function( matched ) {
			return matched && matched.text.length === text.length;
		} )
	} );
} );
[…]
editor.on( 'contentDom', function() {
	var commitKeystrokes = editor.config.autolink_commitKeystrokes || CKEDITOR.config.autolink_commitKeystrokes;
	editor.editable().on( 'keydown', function( evt ) {
		if ( CKEDITOR.tools.indexOf( commitKeystrokes, evt.data.getKey() ) == -1 ) {
			return;
		}
		
		autolink( editor, function( matched ) {
			return matched;
		} );
		// Handle event ASAP thus some plugins may change
		// editor selection or cancel keydown events e.g. wysiwygarea, enterkey.
	}, null, null, 0 );
} );
```

jacekbogdanski commented:
I extracted https://github.com/ckeditor/ckeditor-dev/pull/2197#pullrequestreview-136186428 issues into separate tickets so they will be easy to track and don't block this feature. See #2221 and #2220.

jacekbogdanski commented:
The unit test for continuous typing is disabled for WebKit thus it will result in a false negative. Although I kept manual test for each browser so we will have some safety net for the issue.

jacekbogdanski commented:
I added an additional manual test for filling char sequence Webkit issue. However, I wasn't able to add a unit test for this issue thus Webkit behaves differently when manually typing text inside the editor and when changing editors content using javascript. In manual testing Webkit doesn't extend anchors when typing, however inserting text using javascript changes anchors text node which gives invalid testing results.

The issue doesn't exist in normal feature path, it was caused by special treatment for Firefox and Edge where anchors are extended by typing, so I believe the manual test could be enough for this case.

Comandeer commented on tests/plugins/autolink/manual/continoustyping.html:
```diff
@@ -0,0 +1,5 @@
+<div id="editor"></div>
+
+<script>
+	CKEDITOR.replace( 'editor' );
```
This test should be ignored also in IE as our plugin does not run there.

Comandeer commented on plugins/autolink/plugin.js:
```diff
@@ -6,45 +6,76 @@
+
+			editor.on( 'contentDom', function() {
+				var commitKeystrokes = editor.config.autolink_commitKeystrokes || CKEDITOR.config.autolink_commitKeystrokes;
+				editor.editable().on( 'keydown', function( evt ) {
```
We have dedicated event for key handling in editor, [`key`](https://docs.ckeditor.com/ckeditor4/latest/api/CKEDITOR_editor.html#event-key).

Comandeer commented on plugins/autolink/plugin.js:
```diff
@@ -6,45 +6,76 @@
+						return matched;
+					} );
+
+					// Handle event ASAP thus some plugins may change
```
I'm not sure if we should interfere with other plugins in such way. If something is trying to stop keyboard event propagation, then it has some important reason for doing it.

jacekbogdanski commented:
I fixed most of the issues except Blink one which has been extracted into a separate issue. See #2239 and #2240 for more information.

Comandeer commented:
Rebased onto latest `major`.
